src/charm.py

```diff
@@ -46,25 +49,68 @@ def _on_install(self, _: ops.InstallEvent) -> None:
         init_config = state.BuilderInitConfig.from_charm(self)
         builder.initialize(init_config=init_config)
         self.unit.status = ops.ActiveStatus("Waiting for first image.")
-        self._run()
 
     @charm_utils.block_if_invalid_config(defer=False)
     def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
         """Handle charm configuration change events."""
-        proxy.configure_aproxy(proxy=state.ProxyConfig.from_env())
         init_config = state.BuilderInitConfig.from_charm(self)
+        if not self._is_image_relation_ready_set_status(config=init_config.run_config):
+            return
+        proxy.configure_aproxy(proxy=state.ProxyConfig.from_env())
         builder.install_clouds_yaml(cloud_config=init_config.run_config.cloud_config)
         if builder.configure_cron(unit_name=self.unit.name, interval=init_config.interval):
             self._run()
         self.unit.status = ops.ActiveStatus()
 
     @charm_utils.block_if_invalid_config(defer=False)
-    def _on_run_action(self, _: ops.EventBase) -> None:
-        """Handle the run action event."""
+    def _on_image_relation_changed(self, _: ops.RelationChangedEvent) -> None:
+        """Handle charm image relation changed event."""
+        init_config = state.BuilderInitConfig.from_charm(self)
+        if not self._is_image_relation_ready_set_status(config=init_config.run_config):
```

Comment:
image_relation_changed and config-change are doing the same thing right?
Not a blocker but if this is true, I think would be easier to understand what happens by extracting the action to one single method called by both handlers. WDYT

Reply:
The environment setup (proxy, clouds.yaml) can be factored out but there is a small difference in the image builder run condition. Would it be alright perhaps if the environment setup requires one extra logical step (from proxy and clouds.yaml setup) is added we can think about refactoring that to a new wrapper setup code?

Reply:
Cool, this can be done in a further PR.